### PR TITLE
Mockchain delegation certificates

### PR DESCRIPTION
**Pull request by edolstra**
This extends the mockchain's `Transaction` type with a vector of `Certificate`s, which are signed messages for registering/deregistering a stake key, registering/retiring a stake pool, and delegating from a stake key to a stake pool. They're in `Transaction` to ensure UTxO-based protection against replays and to allow fees/deposits if needed.

Handling of these messages to update the ledger/leadership will be a subsequent PR.

### CODE REVIEW

**Comment by edolstra** [comment]
Updated: Blocks now have a list of `Message`s, which are an enum of `SignedTransaction` and the various delegation-related certificate types.

File: chain-impl-mockchain/src/block.rs [hunk 1]
```diff
@@ -13,7 +14,17 @@ pub struct Block {
     pub slot_id: BlockDate,
     pub parent_hash: Hash,
 
-    pub transactions: Vec<SignedTransaction>,
+    pub messages: Vec<Message>,
```
Comment (NicolasDP) [inline chain-impl-mockchain/src/block.rs]: `block_contents: Vec<Message>` I think we want to highlight that this si the content of the block and everything else is meta.

File: chain-impl-mockchain/src/block.rs [hunk 2]
```diff
@@ -277,19 +286,91 @@ impl property::Deserialize for SignedBlockSummary {
...
+pub const TAG_STAKE_POOL_RETIREMENT: u8 = 6;
+
+impl property::Serialize for Message {
+    type Error = std::io::Error;
+    fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), Self::Error> {
+        use chain_core::packer::*;
+        let mut codec = Codec::from(writer);
+        match self {
+            Message::Transaction(signed) => {
+                codec.put_u8(TAG_TRANSACTION)?;
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/block.rs]: Are tagged values better here, rather than serializing events of different types in separate (possibly type-tagged) arrays?

File: shell.nix [hunk 3]
```diff
@@ -1,4 +1,4 @@
-{ nixpkgs ? fetchTarball channel:nixos-unstable
+{ nixpkgs ? fetchTarball https://github.com/NixOS/nixpkgs/archive/7fff567ee99c1f343ecdd82fef2e35fb6f50e423.tar.gz
```
Comment (mzabaluev) [inline shell.nix]: Ding-ding, unrelated change :)

File: cardano/src/block/block.rs [hunk 4]
```diff
@@ -361,11 +361,11 @@ impl chain_core::property::Deserialize for Block {
 }
 
 impl chain_core::property::HasTransaction for Block {
-    type Transactions = [TxAux];
-    fn transactions(&self) -> &Self::Transactions {
+    type Transaction = TxAux;
+    fn transactions<'a>(&'a self) -> Box<Iterator<Item = &Self::Transaction> + 'a> {
```
Comment (mzabaluev) [inline cardano/src/block/block.rs]: This introduces allocations and dynamic dispatch in iterator loops for the "production" blockchain. I'd like to avoid it if possible.

File: chain-core/src/property.rs [hunk 5]
```diff
@@ -147,13 +147,11 @@ pub trait Transaction: Serialize + Deserialize {
...
-    /// A reference to the `Transactions` type must be convertible to an
-    /// iterator returning references to transaction objects.
-    type Transactions: ?Sized;
+    /// The type of transactions in this block.
+    type Transaction;
 
-    /// Returns a reference that can be used to iterate over transactions in the block.
-    fn transactions(&self) -> &Self::Transactions;
+    /// Returns an iterator over the transactions in the block.
+    fn transactions<'a>(&'a self) -> Box<Iterator<Item = &Self::Transaction> + 'a>;
```
[the remark this answers is not in the document]
Comment (edolstra) [reply chain-core/src/property.rs]: I guess that's another argument for keeping these certificates as transactions (in addition to the UTxO-based replay protection)...

File: chain-impl-mockchain/src/block.rs [hunk 6]
```diff
@@ -3,6 +3,8 @@ use crate::certificate;
 use crate::key::{Hash, PrivateKey, PublicKey, Signature, Signed};
 use crate::transaction::*;
 use chain_core::property;
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/block.rs]: Do you have to use both of these traits for this to work?

Comment (edolstra) [reply chain-impl-mockchain/src/block.rs]: Yeah, I think they're both required.
